File: src/FarmEngine.Core/Tools.cs
using FarmEngine.Schemas;

namespace FarmEngine.Core;

/// <summary>
/// Tool rules — extracted from src/lib/tools.ts, behavior-identical
/// (characterization-tested). Port of tools.ts.
/// </summary>
public static class Tools
{
    /// <summary>TS <c>TOOL_DEFINITIONS[toolType]</c>; unknown types yield <c>undefined</c> (null) like TS.</summary>
    public static ToolDefinition GetToolDefinition(string toolType) =>
        ContentBuiltin.ToolDefinitions.TryGetValue(toolType, out var definition) ? definition : null!;

    public static bool CanUseTool(Item tool, Tile targetTile)
    {
        if (string.IsNullOrEmpty(tool.ToolType)) return false;
        if (!ContentBuiltin.ToolDefinitions.TryGetValue(tool.ToolType, out var definition) || definition == null) return false;
        return definition.ValidTargets.Contains(targetTile.Type);
    }

    public static ToolDefinition? GetToolFromItem(Item item)
    {
        if (string.IsNullOrEmpty(item.ToolType)) return null;
        return ContentBuiltin.ToolDefinitions.TryGetValue(item.ToolType, out var definition) ? definition : null;
    }

    /// <summary>JS falsiness of an optional number (undefined, 0, NaN).</summary>
    private static bool Falsy(double? value) => value is not { } v || v == 0 || double.IsNaN(v);

    public static Item DamageToolDurability(Item tool, double amount = 1)
    {
        if (Falsy(tool.Durability) || Falsy(tool.MaxDurability)) return tool;
        return tool with { Durability = Math.Max(0, tool.Durability!.Value - amount) };
    }

    public static bool IsToolBroken(Item tool)
    {
        // M2 fix: a tool at exactly 0 durability IS broken (the old falsy check
        // meant tools could never break; repair shops make breakage meaningful).
        if (tool.Durability == null || tool.MaxDurability == null) return false;
        return tool.Durability.Value <= 0;
    }

    public static Item RepairTool(Item tool, double? amount = null)
    {
        if (Falsy(tool.Durability) || Falsy(tool.MaxDurability)) return tool;
        var repairAmount = Falsy(amount) ? tool.MaxDurability!.Value : amount!.Value;
        return tool with { Durability = Math.Min(tool.MaxDurability!.Value, tool.Durability!.Value + repairAmount) };
    }
}

File: crates/farm-sim/src/tools.rs
//! Tools (port of `Tools.cs` / tools.ts).

use crate::content_builtin::ToolDefinition;
use crate::schema::{Item, Tile};

pub fn get_tool_definition(tool_type: &str) -> ToolDefinition {
    let _ = tool_type;
    todo!("port Tools.GetToolDefinition")
}

pub fn can_use_tool(tool: &Item, target_tile: &Tile) -> bool {
    let _ = (tool, target_tile);
    todo!("port Tools.CanUseTool")
}

pub fn get_tool_from_item(item: &Item) -> Option<ToolDefinition> {
    let _ = item;
    todo!("port Tools.GetToolFromItem")
}

pub fn damage_tool_durability(tool: &Item, amount: f64) -> Item {
    let _ = (tool, amount);
    todo!("port Tools.DamageToolDurability")
}

pub fn is_tool_broken(tool: &Item) -> bool {
    let _ = tool;
    todo!("port Tools.IsToolBroken")
}

pub fn repair_tool(tool: &Item, amount: Option<f64>) -> Item {
    let _ = (tool, amount);
    todo!("port Tools.RepairTool")
}
